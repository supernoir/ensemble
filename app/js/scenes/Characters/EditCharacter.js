import React from 'react';
import intl from 'react-intl-universal';
import { Link } from 'react-router-dom';
import { Container, Segment, Form, Button, Divider, Header, Breadcrumb } from 'semantic-ui-react';
import Loader from '../../basics/Loader';

export default class EditCharacter extends React.Component {
	constructor() {
		super();
		this.state = {
			firstname  : '',
			middlename : '',
			lastname   : '',
			birthday   : '',
			desc       : '',
			gender     : '',
			origin     : '',
			nationality: '',
			family     : '',
			project    : '',
			series     : '',
			imgurl     : ''
		};
	}

	handleInput = (source, evt) => {
		switch (source) {
			case 'firstname':
				this.setState({ firstname: evt.currentTarget.value });
				break;
			case 'middlename':
				this.setState({ middlename: evt.currentTarget.value });
				break;
			case 'lastname':
				this.setState({ lastname: evt.currentTarget.value });
				break;
			case 'gender':
				this.setState({ gender: evt.currentTarget.value });
				break;
			case 'birthday':
				this.setState({ birthday: evt.currentTarget.value });
				break;
			case 'origin':
				this.setState({ origin: evt.currentTarget.value });
				break;
			case 'project':
				this.setState({ project: evt.currentTarget.value });
				break;
			case 'desc':
				this.setState({ desc: evt.currentTarget.value });
				break;
			default:
				break;
		}
	};

	assembleFullName = (first, middle, last) => {
		try {
			let fullName = first !== void 0 && first.length > 0 ? first + ' ' : '';
			fullName += middle !== void 0 && middle.length > 0 ? middle + ' ' : '';
			fullName += last !== void 0 && last.length > 0 ? last : '';
			return fullName.trim();
		} catch (err) {
			throw err;
		}
	};

	postnewCharacter(evt) {
		evt.preventDefault();

		this.props.editCharacter(this.props.match.params.id, {
			desc       : this.state.desc,
			first_name : this.state.firstname,
			middle_name: this.state.middlename,
			last_name  : this.state.lastname,
			full_name  : this.assembleFullName(this.state.firstname, this.state.middlename, this.state.lastname),
			gender     : this.state.gender,
			birthday   : this.state.birthday,
			origin     : this.state.origin,
			project    : this.state.project
		});

		this.props.addEvent({
			user  : 'testAdmin',
			action: 'edit_character',
			ref   : this.assembleFullName(this.state.firstname, this.state.middlename, this.state.lastname)
		});

		this.props.history.push('/characters');
	}

	componentDidMount() {
		this.assembleFullName();
		this.props.getCharacterById(this.props.match.params.id);
		this.props.getProjects();
	}

	componentWillReceiveProps(nextProps) {
		if (nextProps !== this.props) {
			this.setState({
				first_name : nextProps.firstname,
				middle_name: nextProps.middlename,
				last_name  : nextProps.lastname,
				full_name  : nextProps.full_name,
				gender     : nextProps.gender,
				birthday   : nextProps.birthday,
				origin     : nextProps.origin,
				project    : nextProps.project
			});
		}
	}

	render() {
		return this.props.loading
			? <Loader loading={this.props.loading} />
			:	<Container>
				<Breadcrumb>
					<Breadcrumb.Section link>
						<Link to="/">{intl.get('component.dashboard')}</Link>
					</Breadcrumb.Section>
					<Breadcrumb.Divider />
					<Breadcrumb.Section link>
						<Link to="/characters">{intl.get('entity.characters')}</Link>
					</Breadcrumb.Section>
					<Breadcrumb.Divider />
					<Breadcrumb.Section active>{intl.get('character.action-edit')}</Breadcrumb.Section>
				</Breadcrumb>

				<Segment>
					<Header as="h2">{intl.get('character.action-edit')}</Header>

					<Divider />

					<Form>
						<Form.Field>
							<label for="desc">{intl.get('character.label-desc')}</label>
							<input
								onChange={evt => this.handleInput('desc', evt)}
								defaultValue={this.props.character.desc}
								type="text"
								id="desc"
								placeholder="A most loveable person"
							/>
						</Form.Field>
						<Form.Field>
							<label for="firstname">{intl.get('character.label-firstname')}</label>
							<input
								onChange={evt => this.handleInput('firstname', evt)}
								defaultValue={this.props.character.first_name}
								type="text"
								id="firstname"
								placeholder="Jane"
								required
							/>
						</Form.Field>
						<Form.Field>
							<label for="middlename">
								{intl.get('character.label-middlename')}
							</label>
							<input
								onChange={evt => this.handleInput('middlename', evt)}
								defaultValue={this.props.character.middle_name}
								type="text"
								id="middlename"
								placeholder="Agatha"
							/>
						</Form.Field>
						<Form.Field>
							<label for="lastname">
								{intl.get('character.label-lastname')}
							</label>
							<input
								onChange={evt => this.handleInput('lastname', evt)}
								defaultValue={this.props.character.last_name}
								type="text"
								id="lastname"
								placeholder="Doe"
							/>
						</Form.Field>
						<Form.Field>
							<label for="gender" className="col-sm-2 control-label">
								{intl.get('character.label-gender')}
							</label>
							<input
								onChange={evt => this.handleInput('gender', evt)}
								defaultValue={this.props.character.gender}
								type="text"
								className="form-control"
								id="gender"
								placeholder="Non-binary"
							/>
						</Form.Field>
						<Form.Field>
							<label for="origin" className="col-sm-2 control-label">
								{intl.get('character.label-origin')}
							</label>
							<input
								onChange={evt => this.handleInput('origin', evt)}
								defaultValue={this.props.character.origin}
								type="text"
								className="form-control"
								id="origin"
								placeholder="Wakanda"
							/>
						</Form.Field>
						<Form.Field>
							<label for="birthday" className="col-sm-2 control-label">
								{intl.get('character.label-birthday')}
							</label>
							<input
								onChange={evt => this.handleInput('birthday', evt)}
								defaultValue={this.props.character.birthday}
								type="text"
								className="form-control"
								id="birthday"
								placeholder="01-07-2019"
							/>
						</Form.Field>

						<Divider />

						<Form.Field>
							<label for="project" className="col-sm-2 control-label">
								{intl.get('entity.project')}
							</label>
							<select onChange={evt => this.handleInput('project', evt)}>
								{this.props.projects !== void 0
									? this.props.projects.map(project => {
										return (
											<option selected={this.props.character.project === project.id} value={project._id} key={`charoption-${project._id}`}>
												{project.title}
											</option>
										);
									})
									: null}
							</select>
						</Form.Field>

						<Button onClick={evt => this.postnewCharacter(evt)} type="submit" className="btn btn-default">
							{intl.get('character.action-add')}
						</Button>
					</Form>
				</Segment>
			</Container>;
	}
}
