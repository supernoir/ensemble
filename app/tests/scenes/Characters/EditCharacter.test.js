import React from 'react';
import { shallow } from 'enzyme';

// Import Mocks
import mockCharacter from '../../mocks/mockCharacter';

// Import Components
import EditCharacter from '../../../js/scenes/Characters/EditCharacter';

test('EditCharacter -- Snapshot Test', () => {
	const wrapper = shallow(
		<EditCharacter
			loading={false}
			editCharacter={(id, data) => {
				jest.fn(id, data);
			}}
			getCharacterById={id => jest.fn(id)}
			character={mockCharacter}
			getProjects={() => jest.fn()}
			projects={[]}
			addEvent={data => {
				jest.fn(data);
			}}
			match={{ params: { id: 1 } }}
		/>
	);
	expect(wrapper).toMatchSnapshot();
});
