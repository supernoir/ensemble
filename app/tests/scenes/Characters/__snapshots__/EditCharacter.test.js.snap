// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditCharacter -- Snapshot Test 1`] = `
<Container>
  <Breadcrumb>
    <BreadcrumbSection>
      <Link
        replace={false}
        to="/"
      />
    </BreadcrumbSection>
    <BreadcrumbDivider />
    <BreadcrumbSection>
      <Link
        replace={false}
        to="/characters"
      />
    </BreadcrumbSection>
    <BreadcrumbDivider />
    <BreadcrumbSection
      active={true}
    />
  </Breadcrumb>
  <Segment>
    <Header
      as="h2"
    />
    <Divider />
    <Form
      as="form"
    >
      <FormField>
        <label
          htmlFor="desc"
        />
        <input
          defaultValue="Un ecrivain"
          id="desc"
          onChange={[Function]}
          placeholder="A most loveable person"
          type="text"
        />
      </FormField>
      <FormField>
        <label
          htmlFor="firstname"
        />
        <input
          defaultValue="Michelle"
          id="firstname"
          onChange={[Function]}
          placeholder="Jane"
          required={true}
          type="text"
        />
      </FormField>
      <FormField>
        <label
          htmlFor="middlename"
        />
        <input
          defaultValue="Romaine"
          id="middlename"
          onChange={[Function]}
          placeholder="Agatha"
          type="text"
        />
      </FormField>
      <FormField>
        <label
          htmlFor="lastname"
        />
        <input
          defaultValue="Toulouse-Lautrec"
          id="lastname"
          onChange={[Function]}
          placeholder="Doe"
          type="text"
        />
      </FormField>
      <FormField>
        <label
          className="col-sm-2 control-label"
          htmlFor="gender"
        />
        <input
          className="form-control"
          defaultValue="Non-binary"
          id="gender"
          onChange={[Function]}
          placeholder="Non-binary"
          type="text"
        />
      </FormField>
      <FormField>
        <label
          className="col-sm-2 control-label"
          htmlFor="origin"
        />
        <input
          className="form-control"
          defaultValue="Paris, France"
          id="origin"
          onChange={[Function]}
          placeholder="Wakanda"
          type="text"
        />
      </FormField>
      <FormField>
        <label
          className="col-sm-2 control-label"
          htmlFor="birthday"
        />
        <input
          className="form-control"
          defaultValue="2015-03-03"
          id="birthday"
          onChange={[Function]}
          placeholder="01-07-2019"
          type="text"
        />
      </FormField>
      <Divider />
      <FormField>
        <label
          className="col-sm-2 control-label"
          htmlFor="project"
        />
        <select
          onChange={[Function]}
        />
      </FormField>
      <Button
        as="button"
        className="btn btn-default"
        onClick={[Function]}
        type="submit"
      />
    </Form>
  </Segment>
</Container>
`;
